File: src/emc/usr_intf/axis/scripts/mdi.py
#!/usr/bin/env python
#    This is a component of AXIS, a front-end for LinuxCNC
#    Copyright 2004, 2005, 2006 Jeff Epler <jepler@unpythonic.net>
#
#    This program is free software; you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation; either version 2 of the License, or
#    (at your option) any later version.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with this program; if not, write to the Free Software
#    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

'''Manual Data Input - issue a single line of g-code to the running system

mdi.py may be specified on the commandline, e.g.,
        bin/mdi g0 x0
'''
import sys, os
import linuxcnc

#if len(sys.argv) > 1:
#    linuxcnc.nmlfile = sys.argv[1]
#    del sys.argv[1]

c = linuxcnc.command()
s = linuxcnc.stat()

if len(sys.argv) > 1:
    c.mode(linuxcnc.MODE_MDI)
    c.mdi(" ".join(sys.argv[1:]))
else:
    try:
        while 1:
            mdi = raw_input("MDI> ")
            if mdi == '':
                s.poll()
                print s.position
            else:
                c.mode(linuxcnc.MODE_MDI)
                c.mdi(mdi)
    except (SystemExit, EOFError, KeyboardInterrupt): pass

# vim:sw=4:sts=4:et:

File: src/liblinuxcnc-ui/mdi.cc
//
// Copyright (C) 2014 Sebastian Kuzminsky
//
// This library is free software; you can redistribute it and/or
// modify it under the terms of the GNU Library General Public
// License as published by the Free Software Foundation; either
// version 2 of the License, or (at your option) any later version.
//
// This library is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
// Library General Public License for more details.
//
// You should have received a copy of the GNU Library General Public
// License along with this library; if not, write to the
// Free Software Foundation, Inc., 51 Franklin St, Fifth Floor,
// Boston, MA  02110-1301, USA.
//

#include <stdlib.h>
#include <string.h>

#include "linuxcnc-ui-private.h"


int lui_send_mdi_command(lui_t *lui, const char *mdi) {
    EMC_TASK_PLAN_EXECUTE task_plan_execute_msg;

    if (lui_get_task_mode(lui) != lui_task_mode_mdi) {
        return -1;
    }

    // If the length of the passed-in MDI command plus the terminating NULL
    // is bigger than the size of the NML command message buffer, then we
    // error out here.
    if (strlen(mdi) + 1 > sizeof(task_plan_execute_msg.command)) {
        return -1;
    }

    // It's safe to use strcpy here because we just verified that the
    // command will fit.
    strcpy(task_plan_execute_msg.command, mdi);

    return lui_send_nml_command_and_wait(lui, task_plan_execute_msg);
}
